algebird-core/src/main/scala/com/twitter/algebird/ExpHist.scala

```diff
@@ -0,0 +1,405 @@
+
+    /**
+     * @param s the number to convert to l-canonical form
+     * @param l the "l" in l-canonical form
+     * @return vector of the coefficients of 2^i in the
+     *         l-canonical representation of s.
+     *
+     * For example:
+     *
+     * {{{
+     * scala> Canonical.fromLong(15, 2)
+     * res0: Vector[Int] = Vector(3, 2, 2)
+     * }}}
+     * 15 = (3 * 2^0) + (2 * 2^1) + (2 * 2^2)
+     *
+     *
+     * the "l" in l-canonical means that
+     *
+     *  - all return vector entries but the last one == `l` or `l + 1`
+     *  - 1 <= `returnVector.last` <= l + 1
+     *
+     * ## L-Canonical Representation Procedure:
+     *
+     * - Find the largest j s.t. 2^j <= (s + l) / (1 + l)
+     * - let s' = 2^j(1 + l) - l
+     *
+     * - let diff = (s - s') is the position of s within that group.
+     * - let b = the little-endian binary rep of diff % (2^j - 1)
+     * - let ret = return vector of length j:
+     *
+     * {{{
+     * (0 until j).map { i => ret(i) = b(i) + l }
+     * ret(j) = math.floor(diff / 2^j)
+     * }}}
+     */
+    def fromLong(s: Long, l: Int): CanonicalVector =
+      if (s <= 0) CanonicalVector(Vector.empty)
+      else {
+        val num = s + l
+        val denom = l + 1
+        val j = floorPowerOfTwo(num / denom)
+        val offset = (num - (denom << j)).toInt
+        CanonicalVector(
+          binarize(modPow2Minus1(offset, j), j, l) :+ (quotientPow2(offset, j) + 1))
+      }
+
+    /**
+     * @param s the number to convert to l-canonical form
+     * @param l the "l" in l-canonical form
+     * @return vector of numbers that sum to s. Each
+     *         entry is a power of 2, and the number of entries of
+     *         each power of 2 matches the l-canonical
+     *         representation of s.
+     *
+     * Note that:
+     *
+     * {{{
+     * bucketsFromLong(s, l) == fromLong(s, k).toBuckets
+     * }}}
+     *
+     * bucketsFromLong is more efficient.
+     */
+    def bucketsFromLong(s: Long, l: Int): Vector[Long] =
+      if (s <= 0) Vector.empty
+      else {
+        val num = s + l
+        val denom = l + 1
+        val j = floorPowerOfTwo(num / denom)
+        val offset = (num - (denom << j)).toInt
+        val prefixRep = modPow2Minus1(offset, j)
+
+        (0 until j).toVector.flatMap {
+          idx => Vector.fill(l + bit(prefixRep, idx))(1L << idx)
+        } ++ List.fill(quotientPow2(offset, j) + 1)(1L << j)
+      }
+  }
+
+  case class CanonicalVector(rep: Vector[Int]) extends AnyVal {
+    def sum: Int = rep.sum
+
+    /**
+     * Expands out an l-canonical representation into the original number.
+     *
+     * @param rep l-canonical representation of some number s for some l
+     * @return The original s
+     */
+    def toLong: Long =
+      if (rep.isEmpty) 0L
+      else {
+        rep.iterator.zipWithIndex
+          .map { case (i, exp) => i.toLong << exp }
+          .reduce(_ + _)
```

Comment:
there is going to be a fair amount of boxing cost here. Actually calling `Monoid.sum( )` will avoid some of that since `Monoid` is specialized on `Long`